### PR TITLE
Oauth client assertion support

### PR DESCRIPTION
**Pull request by robbertvanwaveren**
This PR adds support for Kafka clients to obtain an access token by authenticating to the authorization server with client assertion as specified by https://www.rfc-editor.org/rfc/rfc7523 and https://www.rfc-editor.org/rfc/rfc7521. The assertion can be provided by an external mechanism and available as a file on the file system or it can be explicitly set through OAuth configuration before running the Kafka client.

In addition, the static access token or refresh token can now be provided as a file on the local filesystem.

The following new configuration options have been added by this PR:
- `oauth.client.assertion`
- `oauth.client.assertion.location`
- `oauth.client.assertion.type`
- `oauth.refresh.token.location`
- `oauth.access.token.location`

### CODE REVIEW

**Comment by robbertvanwaveren** [comment]
this PR now has DCO signoff and grants maintainer edit rights

**Review by scholzj — COMMENTED** [review]
Can we have the description and the PR name reflect what it does?

**Comment by mstruk** [comment]
I added a better description.

**Review by scholzj — APPROVED** [review]

**Comment by scholzj** [comment]
@mstruk Same as the other PR. The Travis tests do not seem to run here.

File: README.md
Comment (tombentley) [inline README.md:956]:
```suggestion
the exact type of the token must also be communicated to the token endpoint and defaults to `urn:ietf:params:oauth:client-assertion-type:jwt-bearer` (which is specified in RFC-7523). 
```

File: README.md
Comment (tombentley) [inline README.md:936]:
```suggestion
together with one of authentication options below.
```

File: README.md
Comment (tombentley) [inline README.md:960]:
```suggestion
-  `oauth.client.assertion.type` (i.e. use `urn:ietf:params:oauth:client-assertion-type:saml2-bearer`, specified in RFC-7522, for SAML 2 tokens)
```

File: oauth-client/src/main/java/io/strimzi/kafka/oauth/client/JaasClientOauthLoginCallbackHandler.java
Comment (tombentley) [inline oauth-client/src/main/java/io/strimzi/kafka/oauth/client/JaasClientOauthLoginCallbackHandler.java:219]: Do we want to accept the case where both parameters are non-null? It's ambiguous what the user's intent is in this case.

File: oauth-common/src/main/java/io/strimzi/kafka/oauth/common/OAuthAuthenticator.java
Comment (tombentley) [inline oauth-common/src/main/java/io/strimzi/kafka/oauth/common/OAuthAuthenticator.java:191]: Can we put each parameter on a separate line.

File: oauth-common/src/main/java/io/strimzi/kafka/oauth/common/OAuthAuthenticator.java
Comment (tombentley) [inline oauth-common/src/main/java/io/strimzi/kafka/oauth/common/OAuthAuthenticator.java:225]: Can we put each parameter on a separate line.

**Review by tombentley — COMMENTED** [review]
We shouldn't assume that people trying to use this understand what they're doing. Even if they do it's easy to become focussed on "making it work" and forget about all the steps needed to actually make it secure. so I think:

1. the README should recommend the use of filesystem permissions to protect tokens on disk.
2. we should consider, when the OS is unix, checking the file permissions and failing if they token files are readable by all, or the group. Obviously this doesn't provide any protection on non-unix, but it's better than nothing in my opinion.

**Comment by mstruk** [comment]
> 1. the README should recommend the use of filesystem permissions to protect tokens on disk.

Good point, I'll add some docs to address this.

> 2. we should consider, when the OS is unix, checking the file permissions and failing if they token files are readable by all, or the group. Obviously this doesn't provide any protection on non-unix, but it's better than nothing in my opinion.

Failing is a harsh reaction to this. Especially if an external mechanism takes care of mounting a filesystem so it's all automatic and possibly out of your control. I would prefer to log a warning, but not fail. WDYT?

**Comment by tombentley** [comment]
> Failing is a harsh reaction to this. Especially if an external mechanism takes care of mounting a filesystem so it's all automatic and possibly out of your control. I would prefer to log a warning, but not fail. WDYT?

The 'possibly out of your control point' is a good one. So I guess logging a warning is reasonable.

**Comment by mstruk** [comment]
The 'env:ENV_VARIABLE_NAME' feature doesn't seem to add any capability that couldn't be performed before. Also rather than conflating such a feature with the current PR it is better to introduce it separately if at all.

Also, there is a possibility of unexpected behaviour, for example using `oauth.client.secret=\"env:Hfs67vfv.M8a\"` would automatically trigger a lookup of `Hfs67vfv.M8a` env variable which in this case was not the intention.

It is already the case that when a configuration option is resolved the resolution tries several different override possibilities in a specific order as implemented [here](https://github.com/strimzi/strimzi-kafka-oauth/blob/0.14.0/oauth-common/src/main/java/io/strimzi/kafka/oauth/common/Config.java#L143-L190) and documented [here](https://github.com/strimzi/strimzi-kafka-oauth/tree/0.14.0#configuring-the-oauth2). 

First, if the system property exists with the same name it overrides the JAAS configured attribute. Second, if the env variable with the same name exists that is used. Third, if the env variable exists with the name equal to configuration key converted to all caps and underscores then that one is used. Only if there are no overrides via System properties and env vars the config parameter as configured via JAAS is used.

According to these rules using `oauth.client.assertion.location="env:TOKEN_FILE"` can be achieved by setting the following env variable in the client's execution environment: `OAUTH_CLIENT_ASSERTION_LOCATION="$TOKEN_FILE"` rather than set as a JAAS parameter.

**Comment by robbertvanwaveren** [comment]
I don’t see how the logic you refer to performs the substitution of the TOKEN_FILE env variable for its actual content.

The original intent of this PR is support for MS Azure Workload Identity which injects certain env vars into the execution environment. Which is the main (only) practic use-case for this PR.

Note that this means that these env var names are not configurable so these should be piped towards this libs configs parama/vars.

**Comment by mstruk** [comment]
@robbertvanwaveren Can you give an example or provide a failing test?

Substitution happens in your kafka-client.yaml (or whatever you call the yaml file that you use) when setting environment vars for the pod.
Or in your bash startup script used as an entry point for your client.

File: oauth-client/src/main/java/io/strimzi/kafka/oauth/client/JaasClientOauthLoginCallbackHandler.java
Comment (mstruk) [reply oauth-client/src/main/java/io/strimzi/kafka/oauth/client/JaasClientOauthLoginCallbackHandler.java]: I aded a logic that if both are set, the location (file) is used rather than the token, and warning is logged.

File: oauth-common/src/main/java/io/strimzi/kafka/oauth/common/OAuthAuthenticator.java
Comment (mstruk) [reply oauth-common/src/main/java/io/strimzi/kafka/oauth/common/OAuthAuthenticator.java]: I aded logic that if both are set, the location (file) is used rather than the token, and warning is logged.

**Comment by mstruk** [comment]
I'm thinking of renaming the `oauth.*.location` config options to `oauth.*.path`. The reasoning is that `location` is too generic, it can mean a file on the local filesystem or a url or something completely different as one of the values from a set of supported values. By saying `path` it makes it super clear (at least to me) that it is the path to the file on the local filesystem.

I wonder what others think?

**Comment by mstruk** [comment]
Actually that won't work as there are existing previous options that refer to files and are called `oauth.*.location` (e.g. `oauth.ssl.truststore.location`) and I prefer to keep consistency in the naming.

**Comment by mstruk** [comment]
@tombentley Do you think this needs another review (specifically, the file permissions check)?

**Pull request merged**
